knimin/lib/shotgun.py

```diff
@@ -0,0 +1,160 @@
+
+    Returns
+    -------
+    np.array of floats
+        A 2D array of floats
+    """
+    qpcr_concentration = np.power(10,((cp_vals - b) / m)) * dil_factor / 1000
+
+    return(qpcr_concentration)
+
+
+def compute_shotgun_pooling_values_eqvol(sample_concs, total_vol = 60.0):
+    """Computes molar concentration of libraries from qPCR Cp values.
+
+    Returns a 2D array of calculated concentrations, in nanomolar units
+
+    Parameters
+    ----------
+    sample_concs : numpy array of float
+        The concentrations calculated via qPCR (nM)
+    total_vol : float
+        The total volume to pool (uL)
+
+    Returns
+    -------
+    np.array of floats
+        A 2D array of floats
+    """
+    per_sample_vol = (total_vol / sample_concs.size) * 1000.0
+    
+    sample_vols = np.zeros([sample_concs.shape[0], sample_concs.shape[1]]) +
+                  per_sample_vol
+    
+    return(sample_vols)
+
+
+def compute_shotgun_pooling_values_qpcr(sample_concs, sample_fracs = None,
+                                        min_conc = 10, floor_conc = 50,
+                                        total_nmol = .00001):
+    """Computes pooling volumes for samples based on qPCR estimates of
+    nM concentrations (`sample_concs`).
+
+    Reads in qPCR values in nM from output of `compute_qpcr_concentration`.
+    Samples must be above a minimum concentration threshold (`min_conc`,
+    default 10 nM) to be included. Samples above this threshold but below a
+    given floor concentration (`floor_conc`, default 50 nM) will be pooled as
+    if they were at the floor concentration, to avoid overdiluting the pool.
+
+    Samples can be assigned a target molar fraction in the pool by passing a
+    np.array (`sample_fracs`, same shape as `sample_concs`) with fractional
+    values per sample. By default, will aim for equal molar pooling.
+
+    Finally, total pooling size is determined by a target nanomolar quantity
+    (`total_nmol`, default .00001). For a perfect 384 library, in which you
+    had all samples at a concentration of exactly 400 nM and wanted a total
+    volume of 60 uL, this would be 0.000024 nmol. 
+
+    Parameters
+    ----------
+    sample_concs: 2D array of float
+        nM calculated by compute_qpcr_concentration
+    sample_fracs: 2D of float
+        fractional value for each sample (default 1/N)
+    min_conc: float
+        minimum nM concentration to be included in pool
+    floor_conc: float
+        minimum value for pooling for samples above min_conc
+        corresponds to a maximum vol in pool
+    total_nmol : float
+        total number of nM to have in pool
+
+    Returns
+    -------
+    sample_vols: np.array of floats
+        the volumes in nL per each sample pooled
+    """
+
+    if sample_fracs is None:
+        sample_fracs = np.ones(sample_concs.shape) / sample_concs.size
+    
+    # get samples above threshold
+    sample_fracs_pass = sample_fracs.copy()
```

Comment:
this will throw an exception if `sample_fracs is None`. In addition to adding logic to resolve this, can a test be added which explicitly passes in `sample_fracs=None` to ensure this branch is exercised?